### PR TITLE
Guideline cleanup

### PR DESCRIPTION
**Pull request by RickWinter**
Cleaned up various sections of the guidelines.
Added some missing sections.

### CODE REVIEW

**Review by LarryOsterman — COMMENTED** [review]
I'd like to understand where some of the new requirements come from before approving.

File: docs/cpp/implementation.md [hunk 1]
```diff
@@ -174,13 +196,37 @@ In general, our advice to consumers of these libraries is to establish logging i
...
+
+- `<application_id>`: optional application-specific string. May contain a slash, but must not contain a space. The string is supplied by the user of the client library, e.g. "XCopy/10.0.4-Preview"
+- `<package_name>`: client library (distribution) package name as it appears to the developer, replacing slashes with dashes and removing the Azure and cpp indicator.  For example, "azure-security-keyvault-secrets-cpp" would specify "azsdk-cpp-security-keyvault-secrets".
+- `<package_version>`: the version of the package. Note: this is not the version of the service
+- `<platform_info>`: information about the currently executing language runtime and OS, e.g. "MSVC/14.10.0(Windows-10-10.0.19041-SP0)"
+
+{% include requirement/SHOULD id="cpp-http-telemetry-dynamic" %} send additional (dynamic) telemetry information as a semi-colon separated set of key-value types in the `X-MS-AZSDK-Telemetry` header.  For example:
+
+```http
+X-MS-AZSDK-Telemetry: class=BlobClient;method=DownloadFile;blobType=Block
```
[the remark this answers is not in the document]
Comment (RickWinter) [reply docs/cpp/implementation.md:221]: Its the same wording in all the languages. I'll add a link to general if I can find a reasonable section.

File: docs/cpp/implementation.md [hunk 2]
```diff
@@ -174,13 +196,37 @@ In general, our advice to consumers of these libraries is to establish logging i
...
+```
+
+The content of the header is a semi-colon key=value list.  The following keys have specific meaning:
```
[the remark this answers is not in the document]
Comment (RickWinter) [reply docs/cpp/implementation.md:224]: Correct, they don't currently. However its there if they want to use it.

File: docs/cpp/introduction.md [hunk 3]
```diff
@@ -360,11 +349,11 @@ For example, some subclasses add a constructor allowing to create an operation i
 
 ##### Conditional Request Methods
 
-> TODO: This section needs to be driven by code in the Core library.
+{% include requirement/MUST id="cpp-conditional-request-methods-etag-options" %} define ETag-related options e.g., `IfMatch`, `IfNoneMatch`, etc., in the service method options.
 
 ##### Hierarchical Clients
 
-> TODO: This section needs to be driven by code in the Core library.
+{% include requirement/MAY id="cpp-hierarchical-clients-subclients-return" %} return clients from other clients e.g., a `DatabaseClient` from a `CosmosClient`.
```
[the remark this answers is not in the document]
Comment (RickWinter) [reply docs/cpp/introduction.md:356]: Agreed. I outlined how it should be structured. Its not mandatory that clients do it.

**Comment by RickWinter** [comment]
> I'd like to understand where some of the new requirements come from before approving.

They come from guidance that all languages share.

File: docs/cpp/implementation.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (RickWinter) [reply docs/cpp/implementation.md:221]: There is no general link. There verbage is in all the languages

File: docs/cpp/implementation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply docs/cpp/implementation.md:224]: Link to general guidelines for this: https://azure.github.io/azure-sdk/general_azurecore.html#azurecore-http-telemetry-dynamic

File: docs/cpp/introduction.md [hunk 4]
```diff
@@ -92,7 +90,16 @@ There exists a distinction that must be made clear with service clients: not all
 
 {% include requirement/MUST id="cpp-service-client-constructor-minimal" %} provide a minimal constructor that takes only the parameters required to connect to the service.
 
-> TODO: Add service client factory pattern examples for connection strings.
+{% highlight cpp %}
+ExampleClient ExampleClient::CreateFromConnectionString(
```
[the remark this answers is not in the document]
Comment (RickWinter) [reply docs/cpp/introduction.md:111]: Added a simple example and moved the connection string to later. Also added verbage that SDKs should recommend passwordless scenarios.

**Comment by LarryOsterman** [comment]
> > I'd like to understand where some of the new requirements come from before approving.
> 
> They come from guidance that all languages share.

In general, when we have C++ guidelines that come from the core guidelines, I think it's important to include links to the origin of the guidelines. That way it's clear that the guideline is azure-wide rather than being language specific.

And FWIW, it doesn't appear that any languages support the X-MS-AZSDK-Telemetry policy :). No idea why but I couldn't find any examples of this.

**Review by LarryOsterman — APPROVED** [review]

**Pull request merged**